### PR TITLE
Make debug_func respect the MBEDTLS_DEBUG_C configuration macro

### PR DESCRIPTION
**Pull request by SpriteOvO**
## Description
I found that the `rsa_debug` function is compiled into the binary file even though the `MBEDTLS_DEBUG_C` macro definition is commented out, because the function pointer is stored in `mbedtls_rsa_info`.

I think `rsa_debug` seems to be meaningless when `MBEDTLS_DEBUG_C` is commented out.
`debug_func` should respect the `MBEDTLS_DEBUG_C` configuration macro.

## Status
READY

## Requires Backporting
NO

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Changelog updated
- [ ] Backported

### CODE REVIEW

**Comment by hanno-arm** [comment]
Hi @SpriteOvO,

Thank you for your report and for opening a PR!

The underlying issue here is that `MBEDTLS_DEBUG_C` controls debug functionality which is specific to the SSL/TLS module. The documentation merely hints at that since only `ssl_xxx.c` are impacted by `MBEDTLS_DEBUG_C`, but clearly this has to be documented explicitly. As it stands, it is natural to assume that `MBEDTLS_DEBUG_C` would cover all debugging functionality, which as you have observed, is not the case. I will open a GitHub issue for the improvement of the documentation of `MBEDTLS_DEBUG_C`.

In particular, the PK debugging API isn't meant to be controlled via `MBEDTLS_DEBUG_C`. If it shall be possible to disable it, we'd need to introduce a separate configuration option for this. Before concretely suggesting that, I'd hear other team member's opinions, though -- ping @gilles-peskine-arm @mpg. 

Finally, you  might wonder why we can't expand the scope of `MBEDTLS_DEBUG_C` to include the PK debugging interface, and here the primary reason is that long-term we want the Crypto and TLS modules of Mbed TLS to be cleanly separated, which in particular means that we want to avoid functionality shared between them.

In summary, I suggested to close this PR and instead (a) make sure that the documentation of `MBEDTLS_DEBUG_C` is improved, and potentially (b) introduce a configuration option controlling the availability of the PK debugging functionality.

Kind regards,
Hanno

**Comment by gilles-peskine-arm** [comment]
In principle, I agree with the intent of this pull request. The pk debug interface is sparsely documented, and it's clearly there only for the sake of the ssl debug interface. So it would make sense to include pk debug only when ssl debug is included.

`MBEDTLS_DEBUG_C` currently only controls whether libmbedtls include the `debug` module. Consider a configuration with only crypto and not tls: `MBEDTLS_DEBUG_C` is disabled in `config.h`. With this change, `mbedtls_pk_info` no longer has a `debug_func` field. This is an ABI break, but not an API break, since `mbedtls_pk_info` is defined in an internal header.

I'm uneasy at having `MBEDTLS_DEBUG_C` control something that isn't the `debug` module or some module that depends on it. Normally, `MBEDTLS_XXX_C` only controls whether `xxx.c` is included in the build, and code that depends on the xxx functionality: not code that's mostly useful to implement xxx. Our configuration interface is already complex enough, I prefer not to make it even more irregular.

We could introduce a new option `MBEDTLS_PK_DEBUG`. It could go as far as excluding the function `mbedtls_pk_debug` from the build. The only argument I have against that is that each added compilation option adds some complexity to the build and the testing. @SpriteOvO How critical is the code size saving from excluding the debug wrappers?

I think we should revisit this when we redesign the configuration interface, probably in 3.0 or in some later 3.x.

**Comment by SpriteOvO** [comment]
@gilles-peskine-arm My goal is not to reduce the code size, I just want to delete the useless strings. Because these strings will provide a lot of useful information to reverse engineers.

**Comment by gilles-peskine-arm** [comment]
What useful information do these strings convey? There are plenty of other strings and artifacts that would identify the code as Mbed TLS.

**Comment by SpriteOvO** [comment]
Usually, when a reverse engineer sees a piece of pseudo-code, he doesn't immediately know what the code is for, he needs to think and experience. If he sees the string "RSA" here, anyone would know that it is encryption/decryption related code.

I didn't intentionally hide that I was using the mbedtls library, so I hope I haven't been misunderstood.
Thank you!

**Comment by gilles-peskine-arm** [comment]
I'm not very knowledgeable about reverse engineering, but as far as I know, it's fairly easy to identify cryptography code. Most symmetric algorithms only have so many ways to implement the core functionality (and if they call special-purpose hardware instructions it's even more obvious). Bignum arithmetic is a telltale of asymmetric cryptography. Hiding cryptographic code would be a project in itself and goes a lot beyond hiding a few strings.

If you have pk parsing enabled, you need the OID module, anyway, and that module is pretty much all strings. RSA itself uses a few OID and in Mbed TLS it requires the OID module.

The pk module itself includes strings in the `pk_info` structure. The debug strings don't add any information.
